src/ucp/core/ucp_request.h

```diff
@@ -129,7 +129,17 @@ struct ucp_request {
             ucp_send_nbx_callback_t cb;         /* Completion callback */
 
             const ucp_proto_config_t *proto_config; /* Selected protocol for the request */
-            ucp_datatype_iter_t      dt_iter;       /* Send buffer state */
+
+            /* This structure holds all mutable fields, and everything else
+             * except common send/recv fields 'status' and 'flags' is immutable
+             * TODO: rework RMA case where length is used instead of dt.offset */
+            struct {
+                union {
+                    ucp_datatype_iter_t  dt_iter;  /* Send buffer state */
+                    ucp_dt_state_t       dt;       /* Position in the send buffer */
```

Comment:
btw, maybe worth moving it below send proto fields where state resided before

Reply:
> btw, maybe worth moving it below send proto fields where state resided before

we have to decide what is more important for us - protocols or dt iterator

Reply:
IMO not worth optimizing it more, we plan to remove dt_state eventually